paddle/fluid/framework/parallel_executor.cc

```diff
@@ -0,0 +1,356 @@
+          }
+        }
+      }
+    }
+  }
+}
+
+void ParallelExecutor::GenerateVar(OpHandle *op_handle,
+                                   const std::string &each_var_name,
+                                   const platform::Place &place) const {
+  auto &vars = member_->vars_[place][each_var_name];
+  size_t version = vars.size();
+  auto &var = vars[version];
+  var.version_ = version;
+  var.generated_op_ = op_handle;
+  var.name_ = each_var_name;
+  var.place_ = place;
+  op_handle->outputs_.emplace_back(&var);
+}
+
+VarHandle *ParallelExecutor::GetVarHandle(const std::string &each_var_name,
+                                          const platform::Place &place) const {
+  auto &var_holders = member_->vars_[place];
+  auto &var_holder = var_holders[each_var_name];
+  VarHandle *var = nullptr;
+  if (var_holder.empty()) {
+    auto &init_var = var_holder[0];
+    init_var.place_ = place;
+    init_var.name_ = each_var_name;
+    init_var.generated_op_ = nullptr;
+    init_var.version_ = 0;
+    var = &init_var;
+  } else {
+    var = &var_holder.rbegin()->second;
+  }
+  return var;
+}
+
+void ParallelExecutor::BCastParamsToGPUs(
+    const ProgramDesc &startup_program) const {
+  auto *main_scope = member_->local_scopes_[member_->main_place_];
+  for (auto *var_desc : startup_program.Block(0).AllVars()) {
+    if (var_desc->GetType() == proto::VarType::LOD_TENSOR) {
+      auto &main_tensor =
+          main_scope->FindVar(var_desc->Name())->Get<LoDTensor>();
+
+      ncclDataType_t data_type = ToNCCLDataType(main_tensor.type());
+      auto &dims = main_tensor.dims();
+      size_t numel = main_tensor.numel();
+      std::vector<std::pair<void *, const platform::DeviceContext *>> mems;
+      mems.emplace_back(
+          const_cast<void *>(main_tensor.data<void>()),
+          new platform::CUDADeviceContext(
+              boost::get<platform::CUDAPlace>(member_->main_place_)));
+
+      for (auto &pair : member_->local_scopes_) {
+        if (pair.first == member_->main_place_) {
+          continue;
+        }
+
+        auto local_scope = pair.second;
+        auto *t = local_scope->Var(var_desc->Name())->GetMutable<LoDTensor>();
+        t->Resize(dims);
+        mems.emplace_back(t->mutable_data(pair.first, main_tensor.type()),
+                          new platform::CUDADeviceContext(
+                              boost::get<platform::CUDAPlace>(pair.first)));
+      }
+
+      // TODO(yy): Invoke ncclBCast here. mems, numel, data_type. The mems[0]
+      // is the src, rests are dests.
+
+      (void)(data_type);
+      (void)(numel);
+
+      // Free Communication Ctx
+      for (auto &pair : mems) {
+        // Release Communication Ctx
+
+        // FIXME: Store CUDA DevCtx to member. Since NCCL All Reduce will use
+        // this
+        delete pair.second;
+      }
+    }
+  }
+}
+
+std::vector<LoDTensor> ParallelExecutor::Run(
+    const std::vector<std::string> &fetch_tensors) {
```

Comment:
Instantiate Variables here?

https://github.com/PaddlePaddle/Paddle/blob/41894da145841ee00448c9f9ed67b2744714fdf9/paddle/fluid/framework/executor.cc#L276-L305

Reply:
Yes. We need to instantiate variables here. We might extract this routine to a global function.